File: 049. Group Anagrams/groupAnagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    var group = {}
    for (let str of strs) {
        let sort_str = [...str].sort().join('')
        if (sort_str in group) {
            group[sort_str].push(str);
        }
        else {
            group[sort_str] = [str]
        }
    }
    return Object.values(group);
};

// Runtime 196 ms / Memory 53 MB
// debug: https://jsfiddle.net/

File: 049. Group Anagrams/groupAnagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        // group string
        unordered_map<string, vector<string>> group;
        for (string s:strs) {
            string sort_s = s;
            sort(sort_s.begin(), sort_s.end());
            group[sort_s].push_back(s); //auto empty value initial
        }
        // get value
        vector<vector<string>> output;
        for (auto item:group) {
            output.push_back(item.second);
        }

        return output;
    }
};